### PR TITLE
Implements DP-SSE and optimiser-agnostic microbatch DP-SGD + moments accountant

### PR DESCRIPTION
**Pull request by gkaissis**
Summary of changes:
- Implements differentially private secure stats exchange using the Laplacian mechanism. Integrates @jopasserat comment about .clone() vs. .copy()
- Implements differentially private optimisation with any optimiser based on microbatch gradient manipulation (compare Abadi et al.). This has a substantial performance ramification of O(n^2) but is completely independent of Opacus which breaks PySyft by introducing backpropagation hooks which are not implemented. 
- Implements moments accountant privacy tracking (implementation taken nearly verbatim from Opacus)

Open issues (for @a1302z)
- The implementation is now in the revision_scripts/dp_training.py file. We have to either add it to PriMIA or pick the functionality from PriMIA we need to do the experiments for revision. 
- PriMIA originally comes with a rather primitive (pun not intended) implementation of DP-SGD from PyDP which is now deprecated and also incompatible with Syft. We should remove all PyDp stuff from the requirement and the environment.

### CODE REVIEW

File: torchlib/find_config_DP.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+        help="Database file to store results.",
+    )
+    parser.add_argument(
+        "--unencrypted_aggregation",
+        action="store_true",
+        help="Train model without secure aggregation",
+    )
+    cmdln_args = parser.parse_args()
+    try:
+        study = opt.create_study(
```
Comment (jopasserat) [inline torchlib/find_config_DP.py:148]: with a multi-objective problem, why not unleashing the power of Genetic Algorithms!! -> turn it into a `multi_objective` study :)

File: torchlib/find_config_DP.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+            study_name="federated_pneumonia{:s}".format(
+                "_unencrypted" if cmdln_args.unencrypted_aggregation else ""
+            )
+            if cmdln_args.federated
+            else "vanilla_pneumonia",
+            storage=cmdln_args.db_file,
+            load_if_exists=True,
+            direction="maximize",
+            pruner=opt.pruners.NopPruner()
```
Comment (jopasserat) [inline torchlib/find_config_DP.py:157]: curious why you opted not to prune? :thinking:

File: torchlib/find_config_DP.py (hunk 1 above)
Comment (gkaissis) [reply torchlib/find_config_DP.py:148]: You would not believe what I'm currently running...
Screenshot 2020-12-31 at 11 13 03

File: torchlib/find_config_DP.py (hunk 2 above)
Comment (gkaissis) [reply torchlib/find_config_DP.py:157]: We are, it's jut not pushed.
